Tests/Tests/Encryption/Keychain/CDTEncryptionKeychainManagerTests.m

```diff
@@ -0,0 +1,337 @@
+
+    XCTAssertEqualObjects(expectedSteps, highlevelSteps, @"Method did not behave as expected");
+}
+
+@end
+
+@implementation CDTEncryptionCustomKeychainManager
+
+#pragma mark - Init object
+- (instancetype)initWithStorage:(CDTEncryptionKeychainStorage *)storage
+{
+    self = [super initWithStorage:storage];
+    if (self) {
+        _lastSteps = [NSMutableArray array];
+    }
+
+    return self;
+}
+
+#pragma mark - CDTEncryptionKeychainManager methods
+- (BOOL)encryptionKeyDataAlreadyGenerated
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsEncryptionKeyDataAlreadyGenerated];
+
+    return [super encryptionKeyDataAlreadyGenerated];
+}
+
+#pragma mark - CDTEncryptionKeychainManager+Internal methods
+- (BOOL)isValidEncryptionKeyData:(CDTEncryptionKeychainData *)data
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsIsValidEncryptionKeyData];
+
+    return [super isValidEncryptionKeyData:data];
+}
+
+- (NSData *)generateDpk
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsGenerateDpk];
+
+    return [super generateDpk];
+}
+
+- (CDTEncryptionKeychainData *)keychainDataToStoreDpk:(NSData *)dpk
+                                encryptedWithPassword:(NSString *)password
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsKeychainDataToStoreDpk];
+
+    return [super keychainDataToStoreDpk:dpk encryptedWithPassword:password];
+}
+
+- (NSData *)generatePBKDF2Salt
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsGeneratePBKDF2Salt];
+
+    return [super generatePBKDF2Salt];
+}
+
+- (NSData *)generateAESKeyUsingPBKDF2ToDerivePassword:(NSString *)password
+                                             withSalt:(NSData *)salt
+                                           iterations:(NSInteger)iterations
+                                               length:(NSUInteger)length
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsGenerateAESKey];
+
+    return [super generateAESKeyUsingPBKDF2ToDerivePassword:password
+                                                   withSalt:salt
+                                                 iterations:iterations
+                                                     length:length];
+}
+
+- (NSData *)generateAESIv
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsGenerateAESIv];
+
+    return [super generateAESIv];
+}
+
+- (NSData *)encryptDpk:(NSData *)dpk usingAESWithKey:(NSData *)key iv:(NSData *)iv
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsEncryptDpk];
+
+    return [super encryptDpk:dpk usingAESWithKey:key iv:iv];
+}
+
+- (NSData *)decryptCipheredDpk:(NSData *)cipheredDpk usingAESWithKey:(NSData *)key iv:(NSData *)iv
+{
+    [self.lastSteps addObject:kCDTEncryptionKeychainManagerTestsDecryptCipheredDpk];
+
+    return [super decryptCipheredDpk:cipheredDpk usingAESWithKey:key iv:iv];
+}
+
+@end
```

Comment:
Getting quite close here to requiring a proper mock object framework to keep our lines of code down. Think we're just about okay.